src/OrleansRuntime/Async/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.CodeGeneration;
+using Orleans.Runtime;
+using Orleans.Runtime.Providers;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private static readonly Interner<long, GrainCancellationToken> _cancellationTokens;
+        private static readonly TimeSpan _cleanupFrequency = TimeSpan.FromMinutes(3);
+        private static readonly int _defaultInternerCollectionSize = 31;
+
+        static CancellationSourcesExtension()
+        {
+            _cancellationTokens = new Interner<long, GrainCancellationToken>(
+                _defaultInternerCollectionSize,
+                _cleanupFrequency);
+        }
+
+        public Task CancelTokenSource(uint grainId, GrainCancellationToken token)
+        {
+            GrainCancellationToken gct;
+            if (!_cancellationTokens.TryFind(GetCompoundKey(grainId, token.Id), out gct))
```

Comment:
the key to _cancellationTokens interner can be simply the Guid of token.Id.
we allocate new random Guid for each, so no need to mix in the grain id.